### PR TITLE
Performance Metrics API: Refactored the AZ EventLogger interface

### PR DESCRIPTION
**Pull request by lumberyard-employee-dm**
## What does this PR do?
The EventLogger interface supports logging events in the google trace event format
(https://docs.google.com/document/d/1CvAClvFfyA5R-PhYUmn5OOQtYMH4h6I0nSsKchNAySU/)

The following events are supported: Duration Events, Complete Events, Instant Events, Counter Events and Async Events.

The Flow Events, Object Events, Metadata Events, Memory Dump Events, Mark Events and Clock Sync Event and Context Events are not supported.

Recording of C++ stack frames within events are not supported.

Added an implementation of a JSON Trace Event Logger that logs to the Google Trace Event Format.

Added UnitTest and Benchmarks test that verifies the implementation of the JSON Trace Event Logger.

Add implementation of a stack allocator for rapidjson to use to allocate memory from a pre-defined buffer.
This is used to have a `rapidjson::Writer` produce the JSON output for an event instead of hand-rolling a JSON solution.

Signed-off-by: lumberyard-employee-dm <56135373+lumberyard-employee-dm@users.noreply.github.com>

## How was this PR tested?

Added UnitTest which outputs sample test data

```json
[
  {"name":"Duration Event","id":"0","cat":"Test","ph":"B","ts":1664329933375019,"pid":31760,"tid":36036,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Duration Event","id":"0","cat":"Test","ph":"E","ts":1664329933375119,"pid":31760,"tid":36036,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Complete Event","id":"0","cat":"Test","ph":"X","ts":1664329933375182,"pid":31760,"tid":36036,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Instant Event","id":"0","cat":"Test","ph":"i","ts":1664329933375235,"pid":31760,"tid":36036,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Duration Event","id":"1","cat":"Test","ph":"B","ts":1664329933375233,"pid":31760,"tid":1936,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Counter Event","id":"0","cat":"Test","ph":"C","ts":1664329933375286,"pid":31760,"tid":36036,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0},"frameTime":16}},
  {"name":"Duration Event","id":"1","cat":"Test","ph":"E","ts":1664329933375353,"pid":31760,"tid":1936,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Duration Event","id":"2","cat":"Test","ph":"B","ts":1664329933375328,"pid":31760,"tid":33928,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Async Event","id":"0","cat":"Test","ph":"b","ts":1664329933375356,"pid":31760,"tid":36036,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Duration Event","id":"3","cat":"Test","ph":"B","ts":1664329933375355,"pid":31760,"tid":19832,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Complete Event","id":"1","cat":"Test","ph":"X","ts":1664329933375422,"pid":31760,"tid":1936,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Async Instant Event","id":"0","cat":"Test","ph":"n","ts":1664329933375436,"pid":31760,"tid":36036,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Duration Event","id":"2","cat":"Test","ph":"E","ts":1664329933375458,"pid":31760,"tid":33928,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Instant Event","id":"1","cat":"Test","ph":"i","ts":1664329933375475,"pid":31760,"tid":1936,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Duration Event","id":"3","cat":"Test","ph":"E","ts":1664329933375475,"pid":31760,"tid":19832,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Async Event","id":"0","cat":"Test","ph":"e","ts":1664329933375484,"pid":31760,"tid":36036,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Complete Event","id":"2","cat":"Test","ph":"X","ts":1664329933375506,"pid":31760,"tid":33928,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Counter Event","id":"1","cat":"Test","ph":"C","ts":1664329933375525,"pid":31760,"tid":1936,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0},"frameTime":17}},
  {"name":"Complete Event","id":"3","cat":"Test","ph":"X","ts":1664329933375527,"pid":31760,"tid":19832,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Instant Event","id":"2","cat":"Test","ph":"i","ts":1664329933375554,"pid":31760,"tid":33928,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Async Event","id":"1","cat":"Test","ph":"b","ts":1664329933375573,"pid":31760,"tid":1936,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Instant Event","id":"3","cat":"Test","ph":"i","ts":1664329933375575,"pid":31760,"tid":19832,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Counter Event","id":"2","cat":"Test","ph":"C","ts":1664329933375605,"pid":31760,"tid":33928,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0},"frameTime":18}},
  {"name":"Async Instant Event","id":"1","cat":"Test","ph":"n","ts":1664329933375625,"pid":31760,"tid":1936,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Counter Event","id":"3","cat":"Test","ph":"C","ts":1664329933375629,"pid":31760,"tid":19832,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0},"frameTime":19}},
  {"name":"Async Event","id":"2","cat":"Test","ph":"b","ts":1664329933375654,"pid":31760,"tid":33928,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Async Event","id":"1","cat":"Test","ph":"e","ts":1664329933375675,"pid":31760,"tid":1936,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Async Event","id":"3","cat":"Test","ph":"b","ts":1664329933375677,"pid":31760,"tid":19832,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Async Instant Event","id":"2","cat":"Test","ph":"n","ts":1664329933375701,"pid":31760,"tid":33928,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Async Instant Event","id":"3","cat":"Test","ph":"n","ts":1664329933375724,"pid":31760,"tid":19832,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Async Event","id":"2","cat":"Test","ph":"e","ts":1664329933375751,"pid":31760,"tid":33928,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}},
  {"name":"Async Event","id":"3","cat":"Test","ph":"e","ts":1664329933375770,"pid":31760,"tid":19832,"args":{"string":"Hello world","int64_t":-2,"uint64_t":18446462603027808255,"bool":true,"double":64.0,"array":["Hello world",-2,18446462603027808255,true,64.0],"object":{"Field1":"Hello world","Field2":-2,"Field3":18446462603027808255,"Field4":true,"Field5":64.0}}}
]
```

Added benchmark test to just test throughput of writing metrics to an in-memory stream or to a file directly.
```
Run on (64 X 3700 MHz CPU s)
CPU Caches:
  L1 Data 32K (x32)
  L1 Instruction 32K (x32)
  L2 Unified 524K (x32)
  L3 Unified 16777K (x8)
------------------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                                            Time             CPU   Iterations
------------------------------------------------------------------------------------------------------------------------------------------------------
JsonTraceEventLoggerBenchmarkFixture/BM_JsonTraceEventLogger_RecordEventsToInMemoryStream                        30906 ns        30030 ns        21333
JsonTraceEventLoggerBenchmarkFixture/BM_JsonTraceEventLogger_RecordEventsToTemporaryDirUnbufferedFileStream      54867 ns        57199 ns        11200
OKAY AzRunBenchmarks() returned 0
Returning code: 0```

### CODE REVIEW

File: Code/Framework/AzCore/Tests/Metrics/JsonTraceEventLoggerTests.cpp [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+    {
+        // Open the file and truncate the file
+        constexpr AZ::IO::OpenMode openMode = AZ::IO::OpenMode::ModeWrite;
+        const AZ::IO::FixedMaxPath metricsFilepath = m_tempDirectory.GetDirectoryAsFixedMaxPath() / "metrics.json";
+
+        for ([[maybe_unused]] auto _ : state)
+        {
+            RecordMetricsToStream(AZStd::make_unique<AZ::IO::SystemFileStream>(metricsFilepath.c_str(), openMode));
+        }
+    }
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/Tests/Metrics/JsonTraceEventLoggerTests.cpp:482]: Writing metrics to a SystemFileStream is a lot slower than writing to a ByteContainerStream and then performing a single write to a SystemFileStream.

I am thinking of adding a buffered SystemFileStream class that maintains an internal buffer for the user

File: Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h [hunk 2]
```diff
@@ -5,184 +5,193 @@
...
             }
+        private:
+            RapidXMLStreamWriter* m_writer;
+        };
 
-            ~RapidXMLStreamWriter()
-            {
-                FlushCache();
-            }
+        RapidXMLStreamWriter(IO::GenericStream* stream, size_t writeCacheSize = 128 * 1024)
```
Comment (spham-amzn) [inline Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h:53]: nit:
Can we push this default to a constant variable named defaultWriteCacheSize?

File: Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h [hunk 3]
```diff
@@ -5,184 +5,193 @@
...
+            : m_stream(stream)
+            , m_errorCount(0)
+        {
+            m_cache.reserve(writeCacheSize);
+        }
+
+        ~RapidXMLStreamWriter()
```
Comment (spham-amzn) [inline Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h:60]: nit:
If we are not declaring this as a virtual desctrutor, then maybe mark this class ```final```?

File: Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h (hunk 2 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h:53]: The `RapidXMLStreamWriter` class wasn't actually modified by this PR?
This is an identation change

File: Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h [hunk 4]
```diff
@@ -5,184 +5,193 @@
...
+        }
+        size_t Tell() const
+        {
+            AZ_Assert(false, "RapidJSONStreamWriter Tell not supported.");
+            return 0;
+        }
+        char* PutBegin()
+        {
+            AZ_Assert(false, "RapidJSONStreamWriter PutBegin not supported.");
+            return 0;
```
Comment (spham-amzn) [inline Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h:141]:
```suggestion
            return nullptr;
```

File: Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h [hunk 5]
```diff
@@ -5,184 +5,193 @@
...
+        : public RapidJSONWriteStreamNoCache
+    {
+    public:
+        explicit RapidJSONStreamWriter(AZ::IO::GenericStream* stream, size_t writeCacheSize = 128 * 1024)
+            : RapidJSONWriteStreamNoCache(*stream)
+        {
+            m_cache.reserve(writeCacheSize);
+        }
+
+        ~RapidJSONStreamWriter()
```
Comment (spham-amzn) [inline Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h:163]:
```suggestion
        ~RapidJSONStreamWriter() override
```

File: Code/Framework/AzCore/AzCore/Metrics/IEventLogger.h [hunk 6]
```diff
@@ -8,127 +8,329 @@
...
+    //! Event fields structure that can be used to record event argument data
+    //! NOTE: the EventValue does not own any data
+    //! Any string or array data must outlive the EventValue
+    struct EventValue;
+
+    //! Encapsulates an EventField paired with a field name.
+    //! Represents a single json key, value pair field
+    struct EventField;
+
+    //! std::span requires a complete type, so an EventValue pointer and a size member
```
Comment (spham-amzn) [inline Code/Framework/AzCore/AzCore/Metrics/IEventLogger.h:39]: do we want to refer to ```std::span``` or ```AZStd::span``` even if its just an alias?

File: Code/Framework/AzCore/AzCore/Metrics/IEventLogger.h [hunk 7]
```diff
@@ -8,127 +8,329 @@
...
+
+        constexpr void SetArrayValues (AZStd::span<EventValue> arrayValues);
+        constexpr AZStd::span<EventValue> GetArrayValues() const;
+
+    private:
+        EventValue* m_arrayAddress{};
+        size_t m_arraySize{};
+    };
+
+    //! std::span requires a complete type, so an EventField pointer and a size member
```
Comment (spham-amzn) [inline Code/Framework/AzCore/AzCore/Metrics/IEventLogger.h:56]: same comment about AZStd::span here

File: Code/Framework/AzCore/AzCore/Metrics/IEventLogger.h (hunk 6 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/Metrics/IEventLogger.h:39]: Well I always debate which one to mention.
In this case I mean to imply that the `std::span` class which we have a conforming standard implementation in the AZStd namespace is what requires a complete type

File: Code/Framework/AzCore/AzCore/Metrics/JsonTraceEventLogger.cpp [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+
+    using EventJsonWriter = rapidjson::Writer<AZ::IO::RapidJSONWriteStreamNoCache, rapidjson::UTF8<char>, rapidjson::UTF8<char>,
+        AZ::Json::RapidjsonStackAllocator<StackAllocatorSize>>;
+
+    static void AppendNewlineWithIndent(JsonTraceEventLogger::JsonEventString& eventString, int32_t indent, bool prependComma)
+    {
+        if (prependComma)
+        {
+            eventString += ',';
+        }
```
Comment (spham-amzn) [inline Code/Framework/AzCore/AzCore/Metrics/JsonTraceEventLogger.cpp:47]: nit:
You can save an extra '+=' operation by defining an additional 
```
const AZStd::string_view CommaNewLine = ",\n";
```
and change this to
```
eventString += prependComma?CommaNewLine:NewLine;
```

Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/Metrics/JsonTraceEventLogger.cpp:47]: Yeah I'll do that

File: Code/Framework/AzCore/AzCore/Metrics/JsonTraceEventLogger.h [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        : public IEventLogger
+    {
+    public:
+        JsonTraceEventLogger();
+        explicit JsonTraceEventLogger(JsonTraceLoggerEventConfig);
+        //! Generic stream which owned by the JsonEventLogger
+        explicit JsonTraceEventLogger(AZStd::unique_ptr<AZ::IO::GenericStream> stream);
+        JsonTraceEventLogger(AZStd::unique_ptr<AZ::IO::GenericStream> stream, JsonTraceLoggerEventConfig);
+
+        // Json Trace Event Logger is movable, but not copyable
```
Comment (spham-amzn) [inline Code/Framework/AzCore/AzCore/Metrics/JsonTraceEventLogger.h:43]: This comment looks like it belongs to a delete of the copy constructor, not the destructor?

Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/Metrics/JsonTraceEventLogger.h:43]: This comment is actually incorrect.
It is neither copyable nor movable.
I'll delete it

File: Code/Framework/AzCore/Tests/Metrics/JsonTraceEventLoggerTests.cpp [hunk 10]
```diff
@@ -0,0 +1,485 @@
...
+
+            ResultOutcome resultOutcome(AZ::Metrics::IEventLogger::ErrorString("Uninitialized"));
+            {
+                // Record Duration Begin and End Events
+                AZ::Metrics::DurationArgs durationArgs;
+                durationArgs.m_name = "Duration Event";
+                durationArgs.m_cat = "Test";
+                durationArgs.m_args = argsContainer;
+                durationArgs.m_id = idString;
+                resultOutcome = googleTraceLogger.RecordDurationEventBegin(durationArgs);
```
Comment (spham-amzn) [inline Code/Framework/AzCore/Tests/Metrics/JsonTraceEventLoggerTests.cpp:189]: do we want to assert a success outcome for these test record events?

Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/Tests/Metrics/JsonTraceEventLoggerTests.cpp:189]: We EXPECT_TRUE them on the next line.

File: Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h (hunk 2 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h:53]: I'll make the change anyway

File: Code/Framework/AzCore/AzCore/Metrics/IEventLogger.h (hunk 7 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/Metrics/IEventLogger.h:56]: I'll mention both AZStd::span and std::span here

**Review by spham-amzn — APPROVED** [review]

**Review by jeremyong-az — APPROVED** [review]

File: Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h [hunk 11]
```diff
@@ -5,184 +5,192 @@
...
+        AZStd::vector<AZ::u8> m_cache;
+        int m_errorCount{};
+    };
 
-        /**
-         * For use with rapidxml::Writer/PrettyWriter.
-         */
-        class RapidJSONStreamWriter
+    //! Implements the rapidjson::Stream concept
+    class RapidJSONWriteStreamNoCache
```
[the remark this answers is not in the document]
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h:92]: I'll change the class name to `RapidJSONWriteStreamUnbuffered`.

File: Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h [hunk 12]
```diff
@@ -5,184 +5,192 @@
...
-        /**
-         * For use with rapidxml::print.
-         */
-        class RapidXMLStreamWriter
+    public:
+        class PrintIterator final
         {
         public:
-            class PrintIterator
+            PrintIterator(RapidXMLStreamWriter* writer)
```
[the remark this answers is not in the document]
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/IO/TextStreamWriters.h:25]: Might as well make the change while I am in this file.

File: Code/Framework/AzCore/AzCore/JSON/RapidjsonAllocatorAdapter.h [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+
+        static void Free(void*)
+        {
+            // Does not free memory
+        }
+
+    private:
+        alignas(AlignN) AZStd::array<AZStd::byte, SizeN> m_buffer;
+        AZStd::byte* m_freeOffset{};
```
[the remark this answers is not in the document]
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/JSON/RapidjsonAllocatorAdapter.h:95]: Actually our AZStd::byte is std::byte brought into the [AZStd namespace](https://github.com/o3de/o3de/blob/development/Code/Framework/AzCore/AzCore/std/base.h#L34).

That being said the byte class is only being used for reinterpret_cast to a byte sized pointer type.
None of the bitwise operators are being used with it

File: Code/Framework/AzCore/Tests/Metrics/JsonTraceEventLoggerTests.cpp (hunk 1 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/Tests/Metrics/JsonTraceEventLoggerTests.cpp]: Well a lot slower was an overstatement.
Turns out the opening of a file is accounts for most of the slowness of using a SystemFileStream.
The actual writing to the SystemFile is about 55% as fast as writing to an in-memory stream.

I also created a buffered SystemFile stream implementation locally and it turns out it is slower than just directly using SystemFileStream

**Pull request merged**